hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE TypeApplications #-}
+
+-- | Module to deal with time in direct cardano chain layer. Defines the type
+-- for a 'PointInTime' and a means to acquire one via a 'TimeHandle' and
+-- 'queryTimeHandle'.
+module Hydra.Chain.Direct.TimeHandle where
+
+import Hydra.Prelude
+
+import Cardano.Ledger.Alonzo.TxInfo (slotToPOSIXTime)
+import Cardano.Ledger.Babbage.PParams (PParams' (..))
+import Cardano.Slotting.EpochInfo (EpochInfo, hoistEpochInfo)
+import Cardano.Slotting.Slot (SlotNo)
+import Cardano.Slotting.Time (RelativeTime, toRelativeTime)
+import Control.Monad.Trans.Except (runExcept)
+import Hydra.Cardano.Api (
+  CardanoMode,
+  Era,
+  EraHistory (EraHistory),
+  NetworkId,
+  shelleyBasedEra,
+  toLedgerPParams,
+ )
+import Hydra.Chain.CardanoClient (
+  QueryPoint (QueryAt),
+  queryEraHistory,
+  queryProtocolParameters,
+  querySystemStart,
+  queryTip,
+ )
+import qualified Ouroboros.Consensus.HardFork.History as Consensus
+import Ouroboros.Consensus.HardFork.History.Qry (interpretQuery, wallclockToSlot)
+import Plutus.V2.Ledger.Api (POSIXTime)
+
+type PointInTime = (SlotNo, POSIXTime)
+
+data TimeHandle = TimeHandle
+  { -- | Get the current 'PointInTime'
+    currentPointInTime :: Either Text PointInTime
+  , -- | Adjust a 'PointInTime' by some number of slots, positively or
+    -- negatively.
+    adjustPointInTime :: SlotNo -> PointInTime -> Either Text PointInTime
+  }
+
+-- | Compute current Query 'PointInTime' from wall clock and by querying the
+-- node for era history, system start and protocol parameters. "Current" means within somewhere within
+--
+-- NOTE: We use wall clock because the slot in tip is from the last block and
+-- potentially too far in the past.
+queryTimeHandle :: NetworkId -> FilePath -> IO TimeHandle
+queryTimeHandle networkId socketPath = do
+  tip <- queryTip networkId socketPath
+  systemStart <- querySystemStart networkId socketPath (QueryAt tip)
+  eraHistory <- queryEraHistory networkId socketPath (QueryAt tip)
+  pparams <- queryProtocolParameters networkId socketPath (QueryAt tip)
+  relativeTime <- toRelativeTime systemStart <$> getCurrentTime
+  slotNo <- relativeTimeToSlot relativeTime eraHistory
+  let toTime =
+        slotToPOSIXTime
+          (toLedgerPParams (shelleyBasedEra @Era) pparams)
+          (toEpochInfo eraHistory)
+          systemStart
+  pure $
+    TimeHandle
+      { currentPointInTime = (slotNo,) <$> toTime slotNo
+      , adjustPointInTime = \n (slot, _) -> do
+          let adjusted = slot + n
+          time <- toTime adjusted
+          pure (adjusted, time)
+      }
+ where
+  toEpochInfo :: EraHistory CardanoMode -> EpochInfo (Either Text)
+  toEpochInfo (EraHistory _ interpreter) =
+    hoistEpochInfo (first show . runExcept) $
```

Comment:
This is kind of odd now, or more exactly, inconsistent. The interpreter base exception is actually this `PastHorizonException` which we throw in the case of `relativeTimeToSlot` but, we returned shown when converting to POSIX time. I know that this is because of the `slotToPOSIXTime` function API which requires an `EpochInfo (Either Text)` (all hail string-based APIs!) but, I don't know, it is just inconsistent. 

Since we do actually dipped into the interpreter query API, I would be tempted to ditch `slotToPOSIXTime` altogether and run the calculation ourselves using the `Qry` primitives. This would give us more control over the error handling.